### PR TITLE
Clarify mutable objects usage

### PR DESCRIPTION
**Pull request by xclaesse**
Only Environment and ConfigurationData are mutable. However, only ConfigurationData becomes immutable after first use which is inconsistent.

This deprecates modification after first use of Environment object and clarify documentation.

### CODE REVIEW

**Comment by dcbaker** [comment]
I like this change and am in favor of it, so once the CI is green I’m for merging.


I really wish that we required an explicit copy instead of an implicit one, but that ship has sailed :/

**Comment by xclaesse** [comment]
Seems that part is even documented: https://mesonbuild.com/Configuration.html "Copy of immutable configuration_data is still immutable."

**Comment by jpakkane** [comment]
The reason for that is that originally Meson objects were like Java and Python objects in that they only exist "out there" and variables are references. An explicit copy method does make sense, too, because it should be a rare operation.

**Comment by xclaesse** [comment]
I added a FeatureNew if muting a copy of immutable ConfigurationData.

**Comment by dcbaker** [comment]
I don't know how @jpakkane feels, but I'd rather not have more implicit copying instead of referencing. I'd rather have an explicit `.copy()` method for that case, and a suggestion for `>= 1.5` to always use the `.copy()` method in the docs. I'd even be in favor of a 2.0 world of removing the implicit copy-on-assign behavior from objects to get a consistent reference-on-assign behavior for all objects.

**Comment by xclaesse** [comment]
While I agree the implicit copy is not ideal, I think for the time being consistency is more important.

In the following example it does not make much sense that env2 is immutable but not env3. Either we copy on assign or we don't, but current mix is even worse.
```meson
env1 = envirtonment()
env2 = env1
run_command(..., env: env1)
env3 = env1
```

**Comment by dcbaker** [comment]
Okay, that is weird. I may take a swing at some point at removing the copy-on-assign behavior and have an explicit copy instead with proper deprecation of course. For now you can consider this an ack from me

File: mesonbuild/interpreterbase/_unholder.py [hunk 1]
```diff
@@ -5,14 +5,16 @@
...
+from .baseobjects import InterpreterObject, MesonInterpreterObject, ObjectHolder, HoldableTypes, MutableInterpreterObject
 from .exceptions import InvalidArguments
 from ..mesonlib import HoldableObject, MesonBugException
 
 if T.TYPE_CHECKING:
     from .baseobjects import TYPE_var
 
 def _unholder(obj: InterpreterObject) -> TYPE_var:
+    if isinstance(obj, MutableInterpreterObject):
+        obj.mark_used()
```
Comment (eli-schwartz) [inline mesonbuild/interpreterbase/_unholder.py:17]: This is rather excessive. Before this commit we set `obj.used = True` directly in func_configure_file() and also in the cmake module. Which this commit doesn't clean up either. We cannot be more restrictive than that except via a warning...

In retrospect, the fact that this added new ways to mark_used() but didn't delete the old ways was a bit of a red flag that the logic wasn't equivalent.

**Pull request merged**